operator/pkg/lbipam/lbipam.go

```diff
@@ -0,0 +1,1730 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package lbipam
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"strings"
+	"sync"
+	"syscall"
+
+	"github.com/cilium/ipam/service/ipallocator"
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/pflag"
+	"go.uber.org/multierr"
+	"golang.org/x/exp/slices"
+	core_v1 "k8s.io/api/core/v1"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	client_typed_v1 "k8s.io/client-go/kubernetes/typed/core/v1"
+
+	k8sClient "github.com/cilium/cilium/pkg/k8s/client"
+
+	operatorOption "github.com/cilium/cilium/operator/option"
+	"github.com/cilium/cilium/pkg/hive"
+	"github.com/cilium/cilium/pkg/hive/cell"
+	"github.com/cilium/cilium/pkg/k8s"
+	cilium_api_v2alpha1 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2alpha1"
+	cilium_client_v2alpha1 "github.com/cilium/cilium/pkg/k8s/client/clientset/versioned/typed/cilium.io/v2alpha1"
+	"github.com/cilium/cilium/pkg/k8s/resource"
+	slim_meta_v1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/apis/meta/v1"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/stream"
+)
+
+const (
+	// The condition added to services to indicate if a request for IPs could be satisfied or not
+	ciliumSvcRequestSatisfiedCondition = "io.cilium/lb-ipam-request-satisfied"
+
+	// The annotation LB IPAM will look for when searching for requested IPs
+	ciliumSvcLBIPSAnnotation = "io.cilium/lb-ipam-ips"
+
+	// The string used in the FieldManager field on update options
+	ciliumFieldManager = "cilium-operator-lb-ipam"
+
+	serviceNamespaceLabel = "io.kubernetes.service.namespace"
+	serviceNameLabel      = "io.kubernetes.service.name"
+)
+
+var Cell = cell.Module(
+	"LB-IPAM",
+	cell.Config(defaultLBIPAMConfig),
+	// Provide LBIPAM so instances of it can be used while testing
+	cell.Provide(newLBIPAM),
+	// Invoke an empty function which takes an LBIPAM to force its construction.
+	cell.Invoke(func(*LBIPAM) {}),
+)
+
+var defaultLBIPAMConfig = LBIPAMConfig{
+	EnableLBIPAM: false,
+}
+
+type LBIPAMConfig struct {
+	EnableLBIPAM bool `mapstructure:"lb-ipam"`
+}
+
+func (cfg LBIPAMConfig) Flags(flags *pflag.FlagSet) {
+	flags.Bool("lb-ipam", cfg.EnableLBIPAM, "Enables LB IPAM, which manages IPs for LoadBalancer services internally without cloud providers")
```

Comment:
Why make this a flag? If no CRs are present, there's nothing to do...

Reply:
LB-IPAM assumes it is the default IP provider, in case we deploy in k8s <v1.24. Thus, if LB-IPAM is enabled, it will start modifying service statuses even if there are no pools(marking them with conditions explaining that we can't allocate IPs since there are no pools). Hence why it is behind a feature flag

Reply:
Alright, made it so LB-IPAM wakes up when it sees a pool and will start doing its work and goes back to sleep when the last pool has been removed.